src/api/bdaddr.rs

```diff
@@ -0,0 +1,270 @@
+}
+
+impl fmt::Display for BDAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        <Self as fmt::LowerHex>::fmt(self, f)
+    }
+}
+
+impl fmt::LowerHex for BDAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let a = &self.address;
+        write!(
+            f,
+            "{:02x}:{:02x}:{:02x}:{:02x}:{:02x}:{:02x}",
+            a[0], a[1], a[2], a[3], a[4], a[5]
+        )
+    }
+}
+
+impl fmt::UpperHex for BDAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let a = &self.address;
+        write!(
+            f,
+            "{:02X}:{:02X}:{:02X}:{:02X}:{:02X}:{:02X}",
+            a[0], a[1], a[2], a[3], a[4], a[5]
+        )
+    }
+}
+
+impl fmt::Debug for BDAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        <Self as fmt::Display>::fmt(self, f)
+    }
+}
+
+impl AsRef<[u8]> for BDAddr {
+    fn as_ref(&self) -> &[u8] {
+        &self.address
+    }
+}
+
+impl From<[u8; 6]> for BDAddr {
+    /// Build an address from an array.
+    ///
+    /// `address[0]` will be the MSB and `address[5]` the LSB.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use btleplug::api::BDAddr;
+    /// let addr: BDAddr = [0x2a, 0xCC, 0x00, 0x34, 0xfa, 0x00].into();
+    /// assert_eq!("2a:cc:00:34:fa:00", addr.to_string());
+    /// ```
+    fn from(address: [u8; 6]) -> Self {
+        Self { address }
+    }
+}
+
+impl<'a> std::convert::TryFrom<&'a [u8]> for BDAddr {
+    type Error = ParseBDAddrError;
+
+    fn try_from(slice: &'a [u8]) -> Result<Self, Self::Error> {
+        if slice.len() < 6 {
+            Err(ParseBDAddrError::IncorrectByteCount)
+        } else {
+            let mut cpy = [0; 6];
+            cpy.copy_from_slice(&slice[..6]);
+            Ok(cpy.into())
+        }
+    }
+}
+
+impl From<u64> for BDAddr {
+    fn from(int: u64) -> Self {
+        let mut cpy = [0; 6];
+        let slice = int.to_be_bytes(); // Reverse order to have MSB on index 0
+        cpy.copy_from_slice(&slice[2..]);
+        cpy.into()
+    }
+}
+
+impl From<BDAddr> for u64 {
+    fn from(addr: BDAddr) -> Self {
+        let mut slice = [0; 8];
+        (&mut slice[2..]).copy_from_slice(&addr.into_inner());
+        u64::from_be_bytes(slice)
+    }
+}
+
+impl From<ParseBDAddrError> for Error {
+    fn from(e: ParseBDAddrError) -> Self {
+        Error::Other(format!("ParseBDAddrError: {}", e))
+    }
+}
+
+impl FromStr for BDAddr {
+    type Err = ParseBDAddrError;
+
+    /// Parses a Bluetooth address of the form `aa:bb:cc:dd:ee:ff` or of form
+    /// `aabbccddeeff`.
+    ///
+    /// All hex-digits `[0-9a-fA-F]` are allowed.
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.contains(':') {
+            Self::from_str_delim(s)
+        } else {
+            Self::from_str_no_delim(s)
+        }
+    }
+}
+
+impl BDAddr {
+    pub fn into_inner(self) -> [u8; 6] {
+        self.address
+    }
+    pub fn bytes(&self) -> &[u8; 6] {
```

Comment:
This method seems to be the same as `as_ref` above. No need for both.

Reply:
Yeah, if I remember correctly, this was a legacy function but if we are rewriting the whole thing anyway, we can remove it as well.